File: scala/src/main/scala/math/Random.scala
package rays.math

import java.lang.Double.longBitsToDouble

final class Random(var seed: Long):
  def nextLong(): Long =
    seed = (6364136223846793005L * seed) + 1L
    //println(s"Random.nextLong() = ${java.lang.Long.toUnsignedString(seed)}")
    seed

  def nextDoubleBetween(min: Double, max: Double): Double =
    val value: Double = nextLong() & Random.MAX_INT_AS_DOUBLE
    val normalized = value / Random.MAX_INT_AS_DOUBLE
    val res = min + (max - min) * normalized
    //println(s"Random.nextDoubleBetween($min, $max) = $res")
    res

object Random:
  /// Maximum int value that can be represented as a double without loss of
  /// precision.
  val MAX_INT_AS_DOUBLE = 0x0FFFFFFFFFFFFFL

File: swift/Lcg/Lib/Random.swift
public struct Random {
    public var seed: UInt64

    public init(seed: UInt64) {
        self.seed = seed
    }

    /// Maximum int value that can be represented as a double without loss of
    /// precision.
    private static let MAX_INT_AS_DOUBLE: UInt64 = 0x0FFFFFFFFFFFFF

    /// Returns a random `UInt64` value, using the linear congruential generator
    /// algorithm.
    public mutating func nextUInt64() -> UInt64 {
        seed = (6364136223846793005 &* seed) &+ 1
        return seed   
    }

    public mutating func nextDoubleBetween(min: Double, max: Double) -> Double {
        let value = Double(nextUInt64() & Random.MAX_INT_AS_DOUBLE)
        let normalized = value / Double(Random.MAX_INT_AS_DOUBLE)
        return min + (max - min) * normalized
    }   
}
